In [2]:
import holidays
import datetime as dt
from holidays import country_holidays
import pandas as pd
import os
import matplotlib.pyplot as plt

In [3]:
data_folder = 'Data'

df = pd.concat(pd.read_csv(f"{data_folder}/{file}") for file in os.listdir(data_folder))

df.head()

ID                    Date     Primary Type          Description  \
0  1326041.0  01/01/2001 01:00:00 AM          BATTERY               SIMPLE   
1  1319931.0  01/01/2001 01:00:00 PM          BATTERY               SIMPLE   
2  1324743.0  01/01/2001 01:00:00 PM         GAMBLING  ILLEGAL ILL LOTTERY   
3  1310717.0  01/01/2001 01:00:00 AM  CRIMINAL DAMAGE           TO VEHICLE   
4  1318099.0  01/01/2001 01:00:00 AM          BATTERY               SIMPLE   

      Location Description Arrest Domestic    Beat  District  Ward   Latitude  \
0                RESIDENCE  False    False  1624.0      16.0   NaN  41.957850   
1                RESIDENCE  False     True   825.0       8.0   NaN  41.783892   
2                   STREET   True    False   313.0       3.0   NaN  41.780412   
3                   STREET  False    False  2424.0      24.0   NaN  42.012391   
4  RESIDENCE PORCH/HALLWAY  False     True   214.0       2.0   NaN  41.819538   

   Longitude Source URL:   
0 -87.749185          NaN  
1 -87.684841          NaN  
2 -87.611970          NaN  
3 -87.678032          NaN  
4 -87.620020          NaN

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7713119 entries, 0 to 9
Data columns (total 13 columns):
 #   Column                Dtype  
---  ------                -----  
 0   ID                    float64
 1   Date                  object 
 2   Primary Type          object 
 3   Description           object 
 4   Location Description  object 
 5   Arrest                object 
 6   Domestic              object 
 7   Beat                  float64
 8   District              float64
 9   Ward                  float64
 10  Latitude              float64
 11  Longitude             float64
 12  Source URL:           object 
dtypes: float64(6), object(7)
memory usage: 823.8+ MB


In [5]:
missing = df.isna().sum()
percentage = missing * 100 / len(df)

pd.DataFrame({'missing': missing, 'percentage': percentage})

missing  percentage
ID                         10    0.000130
Date                       10    0.000130
Primary Type               10    0.000130
Description                10    0.000130
Location Description    10938    0.141810
Arrest                     10    0.000130
Domestic                   10    0.000130
Beat                       10    0.000130
District                   57    0.000739
Ward                   614856    7.971561
Latitude                88695    1.149924
Longitude               88695    1.149924
Source URL:           7713109   99.999870

In [6]:
df.drop(columns=['ID', 'Latitude', 'Longitude'], inplace=True)
df

Date     Primary Type          Description  \
0   01/01/2001 01:00:00 AM          BATTERY               SIMPLE   
1   01/01/2001 01:00:00 PM          BATTERY               SIMPLE   
2   01/01/2001 01:00:00 PM         GAMBLING  ILLEGAL ILL LOTTERY   
3   01/01/2001 01:00:00 AM  CRIMINAL DAMAGE           TO VEHICLE   
4   01/01/2001 01:00:00 AM          BATTERY               SIMPLE   
..                     ...              ...                  ...   
5                      NaN              NaN                  NaN   
6                      NaN              NaN                  NaN   
7                      NaN              NaN                  NaN   
8                      NaN              NaN                  NaN   
9                      NaN              NaN                  NaN   

       Location Description Arrest Domestic    Beat  District  Ward  \
0                 RESIDENCE  False    False  1624.0      16.0   NaN   
1                 RESIDENCE  False     True   825.0       8.0   NaN   
2                    STREET   True    False   313.0       3.0   NaN   
3                    STREET  False    False  2424.0      24.0   NaN   
4   RESIDENCE PORCH/HALLWAY  False     True   214.0       2.0   NaN   
..                      ...    ...      ...     ...       ...   ...   
5                       NaN    NaN      NaN     NaN       NaN   NaN   
6                       NaN    NaN      NaN     NaN       NaN   NaN   
7                       NaN    NaN      NaN     NaN       NaN   NaN   
8                       NaN    NaN      NaN     NaN       NaN   NaN   
9                       NaN    NaN      NaN     NaN       NaN   NaN   

                                         Source URL:   
0                                                 NaN  
1                                                 NaN  
2                                                 NaN  
3                                                 NaN  
4                                                 NaN  
..                                                ...  
5                                      >> import glob  
6                              >> import pandas as pd  
7                         >> folder = "Data/Chicago/"  
8   >> crime_files = sorted(glob.glob(folder+"*.cs...  
9   >> df = pd.concat([pd.read_csv(f) for f in cri...  

[7713119 rows x 10 columns]

In [7]:
# Dropping rows with missing dates
df.dropna(subset=['Date'], inplace=True)

In [8]:
duplicates = df.duplicated().sum()
duplicates

30401

In [9]:
# Drop duplicates
df.drop_duplicates(inplace=True)

In [10]:
duplicates = df.duplicated().sum()
duplicates

0

In [12]:
date_format = "%m/%d/%Y %H:%M:%S %p"

df['Date'] = pd.to_datetime(df['Date'], format=date_format)
df

Date        Primary Type  \
0      2001-01-01 01:00:00             BATTERY   
1      2001-01-01 01:00:00             BATTERY   
2      2001-01-01 01:00:00            GAMBLING   
3      2001-01-01 01:00:00     CRIMINAL DAMAGE   
4      2001-01-01 01:00:00             BATTERY   
...                    ...                 ...   
238853 2022-12-31 12:50:00             ASSAULT   
238854 2022-12-31 12:50:00             ASSAULT   
238855 2022-12-31 12:52:00             BATTERY   
238856 2022-12-31 12:52:00             ASSAULT   
238857 2022-12-31 12:59:00  DECEPTIVE PRACTICE   

                                    Description          Location Description  \
0                                        SIMPLE                     RESIDENCE   
1                                        SIMPLE                     RESIDENCE   
2                           ILLEGAL ILL LOTTERY                        STREET   
3                                    TO VEHICLE                        STREET   
4                                        SIMPLE       RESIDENCE PORCH/HALLWAY   
...                                         ...                           ...   
238853                                   SIMPLE                     APARTMENT   
238854                     AGGRAVATED - HANDGUN                     APARTMENT   
238855                                   SIMPLE                    BARBERSHOP   
238856                                   SIMPLE  COMMERCIAL / BUSINESS OFFICE   
238857  FINANCIAL IDENTITY THEFT $300 AND UNDER                     RESIDENCE   

       Arrest Domestic    Beat  District  Ward Source URL:   
0       False    False  1624.0      16.0   NaN          NaN  
1       False     True   825.0       8.0   NaN          NaN  
2        True    False   313.0       3.0   NaN          NaN  
3       False    False  2424.0      24.0   NaN          NaN  
4       False     True   214.0       2.0   NaN          NaN  
...       ...      ...     ...       ...   ...          ...  
238853  False    False  1014.0      10.0  24.0          NaN  
238854  False     True   733.0       7.0   6.0          NaN  
238855  False    False  1135.0      11.0  28.0          NaN  
238856  False    False  1432.0      14.0  32.0          NaN  
238857  False    False  1511.0      15.0  29.0          NaN  

[7682708 rows x 10 columns]

In [13]:
df.set_index('Date', inplace=True)

# Topic 1 Comparing Police Districts

    Which district had the most crimes in 2022?
    Which had the least?

In [15]:
data_2022 = df[df.index.year == 2022]
data_2022

Primary Type  \
Date                                      
2022-01-01 01:00:00               THEFT   
2022-01-01 01:00:00  DECEPTIVE PRACTICE   
2022-01-01 01:00:00     CRIMINAL DAMAGE   
2022-01-01 01:00:00       OTHER OFFENSE   
2022-01-01 01:00:00               THEFT   
...                                 ...   
2022-12-31 12:50:00             ASSAULT   
2022-12-31 12:50:00             ASSAULT   
2022-12-31 12:52:00             BATTERY   
2022-12-31 12:52:00             ASSAULT   
2022-12-31 12:59:00  DECEPTIVE PRACTICE   

                                                 Description  \
Date                                                           
2022-01-01 01:00:00                           $500 AND UNDER   
2022-01-01 01:00:00  FINANCIAL IDENTITY THEFT $300 AND UNDER   
2022-01-01 01:00:00                               TO VEHICLE   
2022-01-01 01:00:00                  HARASSMENT BY TELEPHONE   
2022-01-01 01:00:00                           $500 AND UNDER   
...                                                      ...   
2022-12-31 12:50:00                                   SIMPLE   
2022-12-31 12:50:00                     AGGRAVATED - HANDGUN   
2022-12-31 12:52:00                                   SIMPLE   
2022-12-31 12:52:00                                   SIMPLE   
2022-12-31 12:59:00  FINANCIAL IDENTITY THEFT $300 AND UNDER   

                                       Location Description Arrest Domestic  \
Date                                                                          
2022-01-01 01:00:00                              RESTAURANT  False    False   
2022-01-01 01:00:00                                    BANK  False    False   
2022-01-01 01:00:00  PARKING LOT / GARAGE (NON RESIDENTIAL)  False    False   
2022-01-01 01:00:00                               APARTMENT  False    False   
2022-01-01 01:00:00                         OTHER (SPECIFY)  False    False   
...                                                     ...    ...      ...   
2022-12-31 12:50:00                               APARTMENT  False    False   
2022-12-31 12:50:00                               APARTMENT  False     True   
2022-12-31 12:52:00                              BARBERSHOP  False    False   
2022-12-31 12:52:00            COMMERCIAL / BUSINESS OFFICE  False    False   
2022-12-31 12:59:00                               RESIDENCE  False    False   

                       Beat  District  Ward Source URL:   
Date                                                      
2022-01-01 01:00:00   121.0       1.0  42.0          NaN  
2022-01-01 01:00:00   811.0       8.0  23.0          NaN  
2022-01-01 01:00:00   531.0       5.0   9.0          NaN  
2022-01-01 01:00:00   412.0       4.0   8.0          NaN  
2022-01-01 01:00:00  1811.0      18.0   2.0          NaN  
...                     ...       ...   ...          ...  
2022-12-31 12:50:00  1014.0      10.0  24.0          NaN  
2022-12-31 12:50:00   733.0       7.0   6.0          NaN  
2022-12-31 12:52:00  1135.0      11.0  28.0          NaN  
2022-12-31 12:52:00  1432.0      14.0  32.0          NaN  
2022-12-31 12:59:00  1511.0      15.0  29.0          NaN  

[238196 rows x 9 columns]

In [16]:
dis_2022 = data_2022.groupby(['District', 'Arrest']).size().reset_index(name='count')

arrested_2022 = dis_2022[dis_2022['Arrest'] == True].sort_values('count', ascending=False).set_index('District')

total_crimes_2022 = dis_2022.groupby('District')['count'].sum().reset_index(name='total_crimes').sort_values('total_crimes', ascending=False).set_index('District')

arrested_2022

Arrest  count
District               
11.0        True   3091
10.0        True   1688
1.0         True   1645
6.0         True   1581
25.0        True   1546
18.0        True   1454
8.0         True   1381
4.0         True   1298
15.0        True   1277
5.0         True   1259
16.0        True   1246
7.0         True   1222
12.0        True   1168
9.0         True   1126
3.0         True   1085
2.0         True   1049
19.0        True    920
24.0        True    843
14.0        True    835
22.0        True    768
17.0        True    747
20.0        True    451
31.0        True      9

In [17]:
total_crimes_2022

total_crimes
District              
8.0              14770
6.0              14661
12.0             14294
4.0              13904
11.0             13435
1.0              12918
18.0             12391
19.0             12182
25.0             11932
3.0              11915
2.0              11777
9.0              10356
7.0              10310
10.0              9935
5.0               9747
16.0              9196
24.0              8451
15.0              8294
14.0              8134
22.0              7650
17.0              6979
20.0              4950
31.0                15

In [18]:
arrest_rate = (arrested_2022['count'] / total_crimes_2022['total_crimes']) * 100

arrest_rate.sort_values(ascending=False)

District
31.0    60.000000
11.0    23.007071
10.0    16.990438
15.0    15.396672
16.0    13.549369
25.0    12.956755
5.0     12.916795
1.0     12.734169
7.0     11.852570
18.0    11.734323
9.0     10.872924
6.0     10.783712
17.0    10.703539
14.0    10.265552
22.0    10.039216
24.0     9.975151
8.0      9.350034
4.0      9.335443
20.0     9.111111
3.0      9.106169
2.0      8.907192
12.0     8.171261
19.0     7.552126
dtype: float64

### District 8 had the highest crime rate (14770)
### District 31 had the lowest crime rate (15)

>    District 21 had the highest arrest count with 3091, while District 45 had the lowest at only 9 arrests
>
>    District 31 has the highest arrest arrest, at 60% of all crimes commited, while District 19 only arrested about 7.5% of crime perpetrators
> 


# Topic 2) Crimes Across the Years:

    Is the total number of crimes increasing or decreasing across the years?
    Are there any individual crimes that are doing the opposite (e.g., decreasing when overall crime is increasing or vice-versa)?

In [19]:
crimes_per_year = df.resample('Y').size().reset_index(name='Total Crimes')
crimes_per_year['Difference'] = crimes_per_year['Total Crimes'].diff()
# Percentage of increase/decrease from previous year
crimes_per_year['Percentage Change'] = (crimes_per_year['Difference'] / crimes_per_year['Total Crimes']) * 100
total_crimes_2001 = crimes_per_year.loc[crimes_per_year['Date'].dt.year == 2001, 'Total Crimes'].values[0]
crimes_per_year['Change since 2021'] = ((crimes_per_year['Total Crimes'] - total_crimes_2001) / total_crimes_2001) * 100

crimes_per_year

/tmp/ipykernel_394/3079522021.py:1: FutureWarning: 'Y' is deprecated and will be removed in a future version, please use 'YE' instead.
  crimes_per_year = df.resample('Y').size().reset_index(name='Total Crimes')


Date  Total Crimes  Difference  Percentage Change  Change since 2021
0  2001-12-31        484027         NaN                NaN           0.000000
1  2002-12-31        484617       590.0           0.121746           0.121894
2  2003-12-31        473086    -11531.0          -2.437400          -2.260411
3  2004-12-31        466782     -6304.0          -1.350523          -3.562818
4  2005-12-31        451363    -15419.0          -3.416097          -6.748384
5  2006-12-31        445246     -6117.0          -1.373847          -8.012156
6  2007-12-31        434883    -10363.0          -2.382940         -10.153153
7  2008-12-31        425505     -9378.0          -2.203969         -12.090648
8  2009-12-31        391439    -34066.0          -8.702761         -19.128685
9  2010-12-31        369378    -22061.0          -5.972473         -23.686489
10 2011-12-31        350965    -18413.0          -5.246392         -27.490615
11 2012-12-31        335363    -15602.0          -4.652272         -30.713989
12 2013-12-31        306666    -28697.0          -9.357738         -36.642791
13 2014-12-31        275069    -31597.0         -11.486936         -43.170732
14 2015-12-31        264073    -10996.0          -4.164000         -45.442506
15 2016-12-31        268959      4886.0           1.816634         -44.433058
16 2017-12-31        268289      -670.0          -0.249731         -44.571481
17 2018-12-31        268191       -98.0          -0.036541         -44.591727
18 2019-12-31        260652     -7539.0          -2.892362         -46.149285
19 2020-12-31        211688    -48964.0         -23.130267         -56.265250
20 2021-12-31        208271     -3417.0          -1.640651         -56.971202
21 2022-12-31        238196     29925.0          12.563183         -50.788696

In [20]:
crimes_by_type_per_year = df.groupby([df.index.year, 'Primary Type']).size().reset_index(name='Total Crimes')
crimes_by_type_per_year.columns = ['Year', 'Primary Type', 'Total Crimes']

crimes_by_type_per_year

Year                       Primary Type  Total Crimes
0    2001                              ARSON          1010
1    2001                            ASSAULT         31353
2    2001                            BATTERY         93377
3    2001                           BURGLARY         25904
4    2001  CONCEALED CARRY LICENSE VIOLATION             1
..    ...                                ...           ...
684  2022                            ROBBERY          8957
685  2022                        SEX OFFENSE          1225
686  2022                           STALKING           446
687  2022                              THEFT         54525
688  2022                  WEAPONS VIOLATION          8759

[689 rows x 3 columns]

In [21]:
crimes_2001 = crimes_by_type_per_year[crimes_by_type_per_year['Year'] == 2001]
crimes_2022 = crimes_by_type_per_year[crimes_by_type_per_year['Year'] == 2022]
crimes_comparison = pd.merge(crimes_2001, crimes_2022, on='Primary Type', suffixes=('_2001', '_2022'))
crimes_comparison.set_index('Primary Type', inplace=True)
crimes_comparison['Percentage Change'] = ((crimes_comparison['Total Crimes_2022'] - crimes_comparison['Total Crimes_2001']) / crimes_comparison['Total Crimes_2001']) * 100
crimes_comparison.sort_values('Percentage Change', ascending=False)

Year_2001  Total Crimes_2001  Year_2022  \
Primary Type                                                                 
CONCEALED CARRY LICENSE VIOLATION       2001                  1       2022   
CRIMINAL SEXUAL ASSAULT                 2001                 29       2022   
OBSCENITY                               2001                 19       2022   
STALKING                                2001                204       2022   
WEAPONS VIOLATION                       2001               4274       2022   
OTHER NARCOTIC VIOLATION                2001                  6       2022   
DECEPTIVE PRACTICE                      2001              14873       2022   
HOMICIDE                                2001                655       2022   
INTERFERENCE WITH PUBLIC OFFICER        2001                406       2022   
OFFENSE INVOLVING CHILDREN              2001               2258       2022   
MOTOR VEHICLE THEFT                     2001              27480       2022   
ASSAULT                                 2001              31353       2022   
INTIMIDATION                            2001                279       2022   
SEX OFFENSE                             2001               2203       2022   
PUBLIC INDECENCY                        2001                  9       2022   
THEFT                                   2001              98978       2022   
OTHER OFFENSE                           2001              29580       2022   
CRIMINAL DAMAGE                         2001              55464       2022   
ROBBERY                                 2001              18437       2022   
BATTERY                                 2001              93377       2022   
ARSON                                   2001               1010       2022   
CRIMINAL TRESPASS                       2001              13208       2022   
BURGLARY                                2001              25904       2022   
PUBLIC PEACE VIOLATION                  2001               2748       2022   
KIDNAPPING                              2001                933       2022   
LIQUOR LAW VIOLATION                    2001               1634       2022   
NARCOTICS                               2001              50070       2022   
PROSTITUTION                            2001               5914       2022   
GAMBLING                                2001                933       2022   

                                   Total Crimes_2022  Percentage Change  
Primary Type                                                             
CONCEALED CARRY LICENSE VIOLATION                175       17400.000000  
CRIMINAL SEXUAL ASSAULT                         1557        5268.965517  
OBSCENITY                                         47         147.368421  
STALKING                                         446         118.627451  
WEAPONS VIOLATION                               8759         104.936827  
OTHER NARCOTIC VIOLATION                           7          16.666667  
DECEPTIVE PRACTICE                             16373          10.085390  
HOMICIDE                                         713           8.854962  
INTERFERENCE WITH PUBLIC OFFICER                 392          -3.448276  
OFFENSE INVOLVING CHILDREN                      1874         -17.006200  
MOTOR VEHICLE THEFT                            21380         -22.197962  
ASSAULT                                        20777         -33.732019  
INTIMIDATION                                     184         -34.050179  
SEX OFFENSE                                     1225         -44.394008  
PUBLIC INDECENCY                                   5         -44.444444  
THEFT                                          54525         -44.912001  
OTHER OFFENSE                                  14553         -50.801217  
CRIMINAL DAMAGE                                27112         -51.117842  
ROBBERY                                         8957         -51.418344  
BATTERY                                        40890         -

### While gambling and prostitution are almost completely gone, concealed carry license violations have risen by an incredible amount of 17,400%, and criminal sexual assault by 5268%, aswell as some other crimes with a smaller percentage such as Obscenity, stalking and weapons violations doubling from 2001 to 2022

# Topic 3) Comparing AM vs. PM Rush Hour:
    Are crimes more common during AM rush hour or PM rush hour?
    ​What are the top 5 most common crimes during AM rush hour? What are the top 5 most common crimes during PM rush hour?
    Are Motor Vehicle Thefts more common during AM rush hour or PM Rush Hour?

In [29]:
df['AM/PM'] = df.index.strftime('%p')
df['AM/PM'].value_counts()



AM/PM
AM    6803613
PM     879095
Name: count, dtype: int64

In [33]:
percentage_am = (df['AM/PM'] == 'AM').mean() * 100
percentage_am

88.55748519922923

In [39]:
crimes_am = df[(df['AM/PM'] == 'AM')]
crimes_am_grouped = crimes_am.groupby(['Primary Type']).size().reset_index(name='Total Crimes').sort_values('Total Crimes', ascending=False).set_index('Primary Type')

crimes_am_grouped.head(5)

Total Crimes
Primary Type                 
THEFT                 1416763
BATTERY               1276043
CRIMINAL DAMAGE        775414
NARCOTICS              663875
ASSAULT                454707

In [51]:

motor_vehicle_theft = df[df['Primary Type'] == 'MOTOR VEHICLE THEFT']
motor_grouped = motor_vehicle_theft.groupby(['AM/PM']).size().reset_index(name='Total Crimes').set_index('Total Crimes')

motor_grouped

AM/PM
Total Crimes      
326670          AM
38949           PM

    Are crimes more common during AM rush hour or PM rush hour?
They are more common durning AM, 88.5% of crimes are committed at AM rush
    
    ​What are the top 5 most common crimes during AM rush hour? What are the top 5 most common crimes during PM rush hour?
In order they are Theft, Battery, Criminal Damage, Narcotics, Assault

    Are Motor Vehicle Thefts more common during AM rush hour or PM Rush Hour?
They are more common at AM Rush hour, 326,670 cars were stolen at AM rush hour versus 38,949 cars stolen at PM rush hour

# Topic 4) Comparing Months:

    ​What months have the most crime? What months have the least?
    Are there any individual crimes that do not follow this pattern? If so, which crimes?

In [56]:
total_crimes_per_month = (df.resample('M').size()
    .reset_index(name='Total Crimes')
    .assign(Month=lambda x: x['Date'].dt.month)
    .drop(columns='Date')
    .groupby('Month')
    .sum()
    .sort_values(by='Total Crimes', ascending=False)
)
total_crimes_per_month

/tmp/ipykernel_394/3357430459.py:1: FutureWarning: 'M' is deprecated and will be removed in a future version, please use 'ME' instead.
  total_crimes_per_month = (df.resample('M').size()


Total Crimes
Month              
7            714499
8            707366
5            680531
6            679283
10           673169
9            665459
3            627230
4            624956
11           606471
1            598769
12           577620
2            527355

### Most crimes were committed on July and August, while the lowest crimes happened on December and February